google/cloud/storage/examples/storage_bucket_samples.cc

```diff
@@ -221,6 +221,91 @@ void PatchBucketStorageClassWithBuilder(google::cloud::storage::Client client,
   (std::move(client), bucket_name, storage_class);
 }
 
+void AddBucketLabel(google::cloud::storage::Client client, int& argc,
+                    char* argv[]) {
+  if (argc != 4) {
+    throw Usage{"add-bucket-label <bucket-name> <label-key> <label-value>"};
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  auto label_key = ConsumeArg(argc, argv);
+  auto label_value = ConsumeArg(argc, argv);
+  //! [add bucket label] [START storage_add_bucket_label]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string bucket_name, std::string label_key,
+     std::string label_value) {
+    gcs::BucketMetadata updated_metadata = client.PatchBucket(
+        bucket_name,
+        gcs::BucketMetadataPatchBuilder().SetLabel(label_key, label_value));
+    std::cout << "Successfully set label " << label_key << " to " << label_value
+              << " on bucket  " << bucket_name << ".";
+    if (updated_metadata.label_count() == 0U) {
```

Comment:
Is it possible to set a label and then have there be 0 labels?

Reply:
Removed that section.  To answer your question, I do not know, it depends on how the internal concurrency controls in GCS work.  If you had two requests, say `R1: add label`, and `R2: remove label` running concurrently: does GCS guarantee that the returned metadata for `R1` is the state immediately after `R1` executed, or is *a* state that includes **at least** `R1`?

Both are legitimate designs, the documentation does not say (to my knowledge), and probably should not say (in my opinion).

Reply:
Bucket metadata writes are strongly consistent -- if you don't specify a metageneration requirement in the requests that alter metadata, it just depends on which request reaches the backend first as to how your metadata ends up at the end.